spec: move machine type, zone, local ssd defaults out of the TestSpec and the registry

RaduBerinde opened the pull request:
This set of commits makes more progress towards #104029 and - more generally - not baking in any flag configuration into the registry itself.

#### spec: move RoachprodOpts args to separate struct

Epic: none
Release note: None

#### spec: move default machine type from ClusterSpec to RoachprodClusterConfig

This is much more logical and allows the removal of the parameter from
the registry.

Epic: none
Release note: None

#### spec: move default zones to RoachprodClusterConfig

Remove the default zones from `ClusterSpec` (and the registry), and
add it to `RoachprodClusterConfig`.

Epic: none
Release note: None

#### spec: move local SSD preference to RoachprodClusterConfig

We change the boolean in the TestSpec to a tri-state (default, prefer
on, disable). This way we can apply the default when creating the
cluster.

Epic: none
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/111811)
<!-- Reviewable:end -->

renatolabs commented on pkg/cmd/roachtest/spec/cluster_spec.go:
```diff
@@ -59,6 +59,24 @@ func (m MemPerCPU) String() string {
+const (
+	// LocalSSDDefault is the default mode, when the test does not have any
+	// preference. A local SSD may or may not be used, depending on --local-ssd
+	// flag and machine type.
```
Sigh, I wish this was simpler. "Default" could be metamorphic but currently it's defined by `--local-ssd`. The interplay between all of these components (and which one prevails) is pretty arbitrary to me (for example, IMO typically command line flags exist to overrite settings in the code, but in this case, the flag itself is overwritten).

I'd be pretty confused, as a test author. It's hard to reason about when a test is or is not going to use local SSD.

In any case, nothing to change here as this is all existing behaviour; just a rant.

RaduBerinde replied:
And then there is `--roachprod-local-ssd` which actually does override. I agree, it's a mess. I'd be happy to implement if we decide on a desirable and coherent behavior.

We *could* make the `Default` metamorphic unless the `--local-ssd` (or `--local-ssd=false`) flag is passed. But yes I'd want to separate those changes from this refactoring PR.

renatolabs reviewed (approved):
Nice, thanks for doing this!

RaduBerinde committed:
spec: move default zones to RoachprodClusterConfig

Remove the default zones from `ClusterSpec` (and the registry), and
add it to `RoachprodClusterConfig`.

Epic: none
Release note: None

RaduBerinde committed:
spec: move RoachprodOpts args to separate struct

Epic: none
Release note: None

RaduBerinde committed:
spec: move default machine type from ClusterSpec to RoachprodClusterConfig

This is much more logical and allows the removal of the parameter from
the registry.

Epic: none
Release note: None

RaduBerinde committed:
spec: move local SSD preference to RoachprodClusterConfig

We change the boolean in the TestSpec to a tri-state (default, prefer
on, disable). This way we can apply the default when creating the
cluster.

Epic: none
Release note: None

RaduBerinde commented:
bors r+
